File: app/src/main/java/com/wilco375/roosternotification/widget/LesuurWidgetProvider.java
package com.wilco375.roosternotification.widget;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.text.SpannableString;
import android.text.style.StrikethroughSpan;
import android.widget.RemoteViews;

import com.wilco375.roosternotification.R;
import com.wilco375.roosternotification.Schedule;
import com.wilco375.roosternotification.activity.MainActivity;
import com.wilco375.roosternotification.general.ScheduleHandler;
import com.wilco375.roosternotification.general.Utils;

import java.util.Calendar;

public class LesuurWidgetProvider extends AppWidgetProvider {

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        super.onUpdate(context, appWidgetManager, appWidgetIds);

        final int N = appWidgetIds.length;
        for (int i = 0; i < N; i++) {
            int appWidgetId = appWidgetIds[i];

            SharedPreferences sp = context.getSharedPreferences("Main", context.MODE_PRIVATE);

            int dayOfWeek = Calendar.getInstance().get(Calendar.DAY_OF_WEEK);
            if (dayOfWeek == Calendar.SATURDAY || dayOfWeek == Calendar.SUNDAY) return;

            Schedule[] schedule = ScheduleHandler.getScheduleByDay(context, dayOfWeek);
            if (schedule.length < 1) return;

            String subject = "";
            String location = "";
            String timeslot = "";
            boolean cancelled = false;
            //Get 15 minutes before current time
            double currentTime = Utils.calendarTimeToDouble(Calendar.getInstance()) + 0.25;
            for (Schedule lesson : schedule) {
                if (currentTime >= lesson.getStartHour() && currentTime <= lesson.getEndHour()) {
                    subject = lesson.getSubjectAndGroup(sp);
                    location = lesson.getLocation();
                    timeslot = String.valueOf(lesson.getTimeslot());
                    cancelled = lesson.getCancelled();
                }
            }

            Intent intent = new Intent(context, MainActivity.class);
            PendingIntent pendingIntent = PendingIntent.getActivity(context, 0, intent, 0);

            RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.app_widget_lesuur);
            views.setOnClickPendingIntent(R.id.app_widget_lesuur_layout, pendingIntent);
            if (!(subject.equals("") && location.equals("")) && !cancelled) {
                if (!timeslot.equals("0") && !timeslot.equals(""))
                    views.setTextViewText(R.id.app_widget_lesuur_title, timeslot + ": " + subject);
                else views.setTextViewText(R.id.app_widget_lesuur_title, subject);

                views.setTextViewText(R.id.app_widget_lesuur_text, location);
            } else if (cancelled) {
                SpannableString spannableString1;
                if (!timeslot.equals("0") && !timeslot.equals(""))
                    spannableString1 = new SpannableString(timeslot + ": " + subject);
                else spannableString1 = new SpannableString(subject);
                SpannableString spannableString2 = new SpannableString(location);
                spannableString1.setSpan(new StrikethroughSpan(), 0, spannableString1.toString().length(), 0);
                spannableString2.setSpan(new StrikethroughSpan(), 0, spannableString2.toString().length(), 0);
                views.setTextViewText(R.id.app_widget_lesuur_title, spannableString1);
                views.setTextViewText(R.id.app_widget_lesuur_text, spannableString2);
            } else {
                views.setTextViewText(R.id.app_widget_lesuur_title, "");
                views.setTextViewText(R.id.app_widget_lesuur_text, "");
            }

            appWidgetManager.updateAppWidget(appWidgetId, views);
        }
    }
}

File: app/src/main/java/com/wilco375/roosternotification/widget/LesuurWidgetProvider.kt
package com.wilco375.roosternotification.widget

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.Context
import android.content.Intent
import android.view.View
import android.widget.RemoteViews
import com.wilco375.roosternotification.R
import com.wilco375.roosternotification.`object`.Schedule
import com.wilco375.roosternotification.`object`.ScheduleItem
import com.wilco375.roosternotification.activity.MainActivity
import com.wilco375.roosternotification.general.Utils
import io.multimoon.colorful.Colorful
import java.text.SimpleDateFormat
import java.util.*

class LesuurWidgetProvider : AppWidgetProvider() {
    private val hourFormat = SimpleDateFormat("HH:mm", Locale.getDefault())

    override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray) {
        super.onUpdate(context, appWidgetManager, appWidgetIds)

        for (appWidgetId in appWidgetIds) {
            val mainActivityIntent = Intent(context, MainActivity::class.java)
            val pendingIntent = PendingIntent.getActivity(context, 0, mainActivityIntent, 0)
            val views = RemoteViews(context.packageName, R.layout.app_widget_lesuur)

            views.setOnClickPendingIntent(R.id.app_widget_lesdag_layout, pendingIntent)

            val schedule = Schedule.getInstance(context)[Utils.currentScheduleDate()]

            // Get first upcoming lesson with and offset of 15 minutes
            val currentTime = Calendar.getInstance().also { it.add(Calendar.MINUTE, 15) }.time
            val upcomingItem = schedule.getItems().firstOrNull { currentTime <= it.end }

            val sp = Utils.getSharedPreferences(context)
            val teacher = sp.getBoolean("teacher", false)
            val teacherFull = sp.getBoolean("teacherFull", false)
            if (!teacher && !teacherFull) {
                views.setViewVisibility(R.id.app_widget_lesdag_teacher, View.GONE)
            } else {
                views.setViewVisibility(R.id.app_widget_lesdag_teacher, View.VISIBLE)
            }

            val primaryColor = Colorful().getPrimaryColor().getColorPack().normal().asInt()

            views.setImageViewBitmap(R.id.app_widget_lesdag_location_bg,
                    Utils.getRoundedSquareBitmap(50.0, 6.0, primaryColor)
            )

            if (upcomingItem != null) {
                views.setViewVisibility(R.id.has_content, View.VISIBLE)
                views.setViewVisibility(R.id.no_content, View.GONE)

                var summary = upcomingItem.getSubjectAndGroup(sp)
                if (upcomingItem.timeslot != 0) {
                    summary = "${upcomingItem.timeslot}. " + summary
                }
                if (upcomingItem.type != "Les") summary += " (${upcomingItem.type})"
                views.setTextViewText(R.id.app_widget_lesdag_subject, summary)

                if (teacherFull) {
                    views.setTextViewText(R.id.app_widget_lesdag_teacher, upcomingItem.teacherFull)
                } else if (teacher) {
                    views.setTextViewText(R.id.app_widget_lesdag_teacher, upcomingItem.teacher)
                }
                views.setTextViewText(R.id.app_widget_lesdag_time,
                        "${hourFormat.format(upcomingItem.start)} - ${hourFormat.format(upcomingItem.end)}")
                views.setTextViewText(R.id.app_widget_lesdag_location, upcomingItem.location)
            } else {
                views.setViewVisibility(R.id.has_content, View.GONE)
                views.setViewVisibility(R.id.no_content, View.VISIBLE)
            }

            appWidgetManager.updateAppWidget(appWidgetId, views)
        }
    }
}
